### ISSUE TITLE
Transform on canvas is spuriously lost

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#571]
The program at https://code.world/#PRheJxUqzHjsy1I59l3pFhA acts oddly.

It should draw an approximate circle out of random triangles.  Instead, at least on Chrome 62, part of the circle is moved to the top-left corner of the screen.  It's almost as if the Context2D transform applied at the start of the drawing is forgotten partway through.

![image](https://user-images.githubusercontent.com/544744/33791176-e8cd31fc-dc3c-11e7-911a-25e72d8cf13d.png)

**Comment by alphalambda** [comment]
This looks OK in my browser (Firefox 52.4.0 in Linux)

Maybe it is a Chrome bug?

**Comment by cdsmith** [comment]
Yep, confirmed that this works in Firefox (57.0.1 for me), but *not* in Chrome (62.0)

**Comment by cdsmith** [comment]
Confirmed with someone else on this program: https://code.world/#PFBMIUYyWKbQh7NmkTjUj1Q

Interestingly, the entire picture is drawn in place, as well as the shifted version.  So it seems to not be that a time slice expires during drawing, or something like that.  But rather than some event is delivered that repeats the drawing without the transform?

**Comment by chelseaharper** [comment]
I see this in the same program. Got this when program crashed while trying to draw second box

    uncaught exception in Haskell thread: h$ap_1_1_fast: unexpected closure type: 2

**Comment by cdsmith** [comment]
A stack trace from me.

    uncaught exception in Haskell thread: TypeError: Cannot read property 'a' of undefined
    TypeError: Cannot read property 'a' of undefined
        at h$$Al (https://code.world/runJS?mode=codeworld&dhash=DWwwDxSIn0DG_YeyYBMuAMg:47043:11)
        at h$runThreadSlice (https://code.world/runJS?mode=codeworld&dhash=DWwwDxSIn0DG_YeyYBMuAMg:7999:11)
        at h$runThreadSliceCatch (https://code.world/runJS?mode=codeworld&dhash=DWwwDxSIn0DG_YeyYBMuAMg:7957:12)
        at h$mainLoop (https://code.world/runJS?mode=codeworld&dhash=DWwwDxSIn0DG_YeyYBMuAMg:7952:9)
        at https://code.world/runJS?mode=codeworld&dhash=DWwwDxSIn0DG_YeyYBMuAMg:2456:25
        at runIfPresent (https://code.world/runJS?mode=codeworld&dhash=DWwwDxSIn0DG_YeyYBMuAMg:2474:21)
        at onGlobalMessage (https://code.world/runJS?
mode=codeworld&dhash=DWwwDxSIn0DG_YeyYBMuAMg:2508:17)

**Comment by cdsmith** [comment]
It looks like the description of this bug is wrong.  The full image is painted to the screen correctly.  Then, only on Chrome on slow enough computers, the partial image is rendered again at the top corner at a delay.

**Comment by cdsmith** [comment]
Ugh, this is a bigger deal all of a sudden.  Tried to teach fractals in class yesterday, and this kept happening to everyone. :(

**Comment by cdsmith** [comment]
I think I know what's going on here.

In 20b403cd8d19b958dde25de15d089f6d14bd710f, I changed the canvas to stay hidden until a CodeWorld graphics entry point is called.  (This makes console-only programs in code.world/haskell look nicer.)  The result is that (only on some browsers, maybe?) a resize event is delivered in the JavaScript layer while `drawingOf` is running.  If the initial drawing completes in one time slice, then the picture is drawn twice, but there's no big problem.  However, if the initial drawing is preempted, the preemption allows the event to be delivered, and this restarts drawing the same picture in a new Haskell thread.  The two threads proceed to stomp on each other's drawing state, causing the problem seen here.

As a workaround, I've added a `withoutPreemption` to the drawing code, so that it won't be preempted (unless it legitimately blocks, e.g. on an MVar or IO).  This could make it harder to recover from situations like infinite recursion, so it's not ideal; but it's better than being entirely broken.
